association-backend/src/webhooks/handlers/pretix/pretix_event_order_paid.py

```diff
@@ -0,0 +1,171 @@
+
+from src.association.settings import SERVICE_TO_SERVICE_SECRET, USERS_SERVICE_URL
+from src.association_membership.domain.entities import PaymentStatus, PretixPayment
+from src.association_membership.domain.repository import AssociationMembershipRepository
+from src.webhooks.exceptions import NoUserFoundWithEmail
+
+from .api import PretixAPI
+
+logger = logging.getLogger(__file__)
+
+GET_USER_ID_BY_EMAIL = """query UserIdByEmail($email: String!) {
+    userByEmail(email: $email) {
+        id
+    }
+}"""
+
+
+async def pretix_event_order_paid(payload):
+    action = payload["action"]
+    organizer = payload["organizer"]
+    event = payload["event"]
+    order_code = payload["code"]
+
+    pretix_api = PretixAPI(organizer, event)
+
+    order_data = pretix_api.get_order_data(order_code)
+    categories = pretix_api.get_categories()["results"]
+    association_category = next(
+        (
+            category
+            for category in categories
+            if category["internal_name"] == "Association"
+        ),
+        None,
+    )
+
+    if not association_category:
+        logger.info(
+            "Ignoring order_code=%s paid event for organizer=%s event=%s "
+            "because there isn't an association category",
+            order_code,
+            organizer,
+            event,
+        )
+        return
+
+    valid_items = pretix_api.get_items(
+        qs={"category": association_category["id"], "active": "true"}
+    )
+    valid_items_ids = [item["id"] for item in valid_items["results"]]
+
+    order_positions = order_data["positions"]
+    membership_positions = []
+    for position in order_positions:
+        if position["item"] in valid_items_ids:
+            membership_positions.append(position)
+
+    if not membership_positions:
+        logger.info(
+            "No membership positions for order_code=%s so nothing to do", order_code
+        )
+        return
+
+    if len(membership_positions) > 1:
+        raise ValueError(
+            f"Multiple positions found in order_code={order_code} that subscribe the user to the association. This is not supported."
+        )
+
+    user_email = order_data["email"]
+    client = ServiceClient(
+        url=f"{USERS_SERVICE_URL}/internal-api",
+        service_name="users-backend",
+        caller="association-backend",
+        jwt_secret=str(SERVICE_TO_SERVICE_SECRET),
+    )
+    result = await client.execute(GET_USER_ID_BY_EMAIL, {"email": user_email})
+    data = result.data
+
+    if not data["userByEmail"]:
+        raise NoUserFoundWithEmail(
+            f"No user found with the email of order_code={order_code}"
+        )
+
+    user_id = int(data["userByEmail"]["id"])
+    repository = AssociationMembershipRepository()
+
+    idempotency_key = PretixPayment.generate_idempotency_key(
+        organizer, event, order_code
+    )
+    if await repository.is_payment_already_processed(idempotency_key):
```

Comment:
maybe in future we can move the key generation somewhere else, right now we have it in two places (https://github.com/pythonitalia/pycon/pull/2557/files#diff-e0833ab93d303d211218ab08e427cd0d33b122aa332d261b6ee5451e99f72918R86)

Reply:
For now, it should be fine, and unless we want to make our entity use the repository, it isn't something we can easily fix 🤔  unless we let the DB raise an error when trying to save and not do any check before.

> maybe in future we can move the key generation somewhere else

The key itself is only built by `generate_idempotency_key` and then used in the two places (check if the payment already exists and the function where the payment is stored).

If we are worried about `add_payment` change and use a different key, we could pass it as argument